File: packages/ai-gateway/src/adapters/gemini.ts
// Adapter Gemini (REST de Google: generateContent con ?key=).
// Internamente "gemini-flash" se traduce a un id real de la API v1beta.

import { postJson, ProviderHttpError } from "./http.js";
import type { ChatRequest, ChatResult, ModelProvider } from "../types.js";

const MODEL_ALIASES: Record<string, string> = {
  "gemini-flash": "gemini-2.5-flash",
};

export class GeminiProvider implements ModelProvider {
  readonly id = "gemini" as const;

  listModels(): string[] {
    return ["gemini-flash"];
  }

  async chat(req: ChatRequest): Promise<ChatResult> {
    const apiKey = req.apiKey ?? process.env.GEMINI_API_KEY;
    if (!apiKey) throw new Error("Sin API key para gemini (setea la tuya o usa BYOK)");

    const model = MODEL_ALIASES[req.model] ?? req.model;

    const system = req.messages.filter((m) => m.role === "system").map((m) => m.content).join("\n");
    const contents = req.messages
      .filter((m) => m.role !== "system")
      .map((m) => ({ role: m.role === "assistant" ? "model" : "user", parts: [{ text: m.content }] }));

    const data = await postJson<{
      candidates?: Array<{ content?: { parts?: Array<{ text?: string }> } }>;
      usageMetadata?: { promptTokenCount?: number; candidatesTokenCount?: number };
      error?: { message?: string };
    }>({
      url: `https://generativelanguage.googleapis.com/v1beta/models/${encodeURIComponent(model)}:generateContent?key=${encodeURIComponent(apiKey)}`,
      body: {
        contents,
        ...(system ? { systemInstruction: { parts: [{ text: system }] } } : {}),
        generationConfig: {
          temperature: req.temperature ?? 0.7,
          ...(req.maxTokens ? { maxOutputTokens: req.maxTokens } : {}),
        },
      },
      headers: { "x-provider": "gemini" },
      signal: req.signal,
    });

    const content = (data.candidates?.[0]?.content?.parts ?? [])
      .map((p) => p.text ?? "")
      .join("\n");

    if (!content) throw new ProviderHttpError("Respuesta vacía de gemini", 502, "gemini");

    return {
      content,
      model: req.model,
      provider: "gemini",
      usage: {
        inputTokens: data.usageMetadata?.promptTokenCount ?? 0,
        outputTokens: data.usageMetadata?.candidatesTokenCount ?? 0,
      },
    };
  }
}

File: apps/api/internal/gateway/adapters/gemini.go
package adapters

import (
	"fmt"
	"net/url"
	"os"
	"strings"

	"github.com/blu-ia/api/internal/gateway"
)

var modelAliases = map[string]string{
	"gemini-flash": "gemini-2.5-flash",
}

type GeminiProvider struct{}

func (p *GeminiProvider) ID() gateway.ProviderID { return "gemini" }
func (p *GeminiProvider) ListModels() []string   { return []string{"gemini-flash"} }

func (p *GeminiProvider) Chat(req gateway.ChatRequest) (gateway.ChatResult, error) {
	apiKey := ""
	if req.APIKey != nil {
		apiKey = *req.APIKey
	} else {
		apiKey = os.Getenv("GEMINI_API_KEY")
	}
	if apiKey == "" {
		return gateway.ChatResult{}, fmt.Errorf("Sin API key para gemini (setea la tuya o usa BYOK)")
	}
	model := req.Model
	if alias, ok := modelAliases[req.Model]; ok {
		model = alias
	}
	var systemParts []string
	var contents []map[string]any
	for _, m := range req.Messages {
		if m.Role == "system" {
			systemParts = append(systemParts, m.Content)
		} else {
			role := "user"
			if m.Role == "assistant" {
				role = "model"
			}
			contents = append(contents, map[string]any{
				"role":  role,
				"parts": []map[string]string{{"text": m.Content}},
			})
		}
	}
	body := map[string]any{
		"contents": contents,
		"generationConfig": map[string]any{
			"temperature": func() float64 {
				if req.Temperature != nil {
					return *req.Temperature
				}
				return 0.7
			}(),
		},
	}
	if req.MaxTokens != nil {
		body["generationConfig"].(map[string]any)["maxOutputTokens"] = *req.MaxTokens
	}
	if len(systemParts) > 0 {
		body["systemInstruction"] = map[string]any{"parts": []map[string]string{{"text": strings.Join(systemParts, "\n")}}}
	}
	type respBody struct {
		Candidates []struct {
			Content *struct {
				Parts []struct {
					Text *string `json:"text"`
				} `json:"parts"`
			} `json:"content"`
		} `json:"candidates"`
		UsageMetadata *struct {
			PromptTokenCount     *int `json:"promptTokenCount"`
			CandidatesTokenCount *int `json:"candidatesTokenCount"`
		} `json:"usageMetadata"`
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	urlStr := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/%s:generateContent?key=%s", url.PathEscape(model), url.QueryEscape(apiKey))
	data, err := PostJSON[respBody](PostJsonOptions{
		URL:     urlStr,
		Body:    body,
		Headers: map[string]string{"x-provider": "gemini"},
		Ctx:     req.Ctx,
	})
	if err != nil {
		return gateway.ChatResult{}, err
	}
	var sb strings.Builder
	if len(data.Candidates) > 0 && data.Candidates[0].Content != nil {
		for _, p := range data.Candidates[0].Content.Parts {
			if p.Text != nil {
				sb.WriteString(*p.Text)
				sb.WriteString("\n")
			}
		}
	}
	content := strings.TrimSpace(sb.String())
	if content == "" {
		return gateway.ChatResult{}, &ProviderHttpError{Message: "Respuesta vacía de gemini", Status: 502, ProviderLabel: "gemini"}
	}
	var inTok, outTok int
	if data.UsageMetadata != nil {
		if data.UsageMetadata.PromptTokenCount != nil {
			inTok = *data.UsageMetadata.PromptTokenCount
		}
		if data.UsageMetadata.CandidatesTokenCount != nil {
			outTok = *data.UsageMetadata.CandidatesTokenCount
		}
	}
	return gateway.ChatResult{
		Content:  content,
		Model:    req.Model,
		Provider: "gemini",
		Usage:    struct{ InputTokens int `json:"inputTokens"`; OutputTokens int `json:"outputTokens"` }{InputTokens: inTok, OutputTokens: outTok},
	}, nil
}
